src/main/java/org/generationcp/middleware/dao/GermplasmSearchDAO.java

```diff
@@ -1521,4 +1530,108 @@ private boolean addPreFilteredGids(final GermplasmSearchRequest germplasmSearchR
 			.setParameterList("gids", gids) //
 			.list();
 	}
+
+	public List<UserDefinedField> getGermplasmAttributeTypes(final GermplasmSearchRequest germplasmSearchRequest) {
+		try {
+
+			final List<Integer> gids = this.retrieveSearchGids(germplasmSearchRequest, null, null);
+			final String sql = "select distinct {u.*} from atributs a inner join udflds u "
+				+ " where a.atype = u.fldno"
+				+ " and a.gid in (:gids)"
+				+ " order by u.fname";
+
+			final SQLQuery query = this.getSession().createSQLQuery(sql);
+			query.addEntity("u", UserDefinedField.class);
+			query.setParameterList("gids", gids);
+
+			return query.list();
+		} catch (final HibernateException e) {
+			throw new MiddlewareQueryException("Error at getGermplasmAttributeTypes() in GermplasmSearchDAO: " + e.getMessage(), e);
+		}
+	}
+
+	public List<UserDefinedField> getGermplasmNameTypes(final GermplasmSearchRequest germplasmSearchRequest) {
+		try {
+
+			final List<Integer> gids = this.retrieveSearchGids(germplasmSearchRequest, null, null);
+			final String sql = "select distinct {u.*} from names n inner join udflds u "
+				+ " where n.ntype = u.fldno"
+				+ " and n.gid in (:gids)"
+				+ " order by u.fname";
+
+			final SQLQuery query = this.getSession().createSQLQuery(sql);
+			query.addEntity("u", UserDefinedField.class);
+			query.setParameterList("gids", gids);
+
+			return query.list();
+		} catch (final HibernateException e) {
+			throw new MiddlewareQueryException("Error at getGermplasmNameTypes() in GermplasmSearchDAO: " + e.getMessage(), e);
+		}
+	}
+
+	public Map<Integer, Map<Integer, String>> getGermplasmAttributeValues(final GermplasmSearchRequest germplasmSearchRequest) {
```

Comment:
I think it will more reusable if you return List<Attribute> at the Dao level and then convert to a Map (or any other needed structure) at the service level